### DISCUSSION TITLE
Rule parameters - What are they?

### DISCUSSION THREAD

**Discussion by tfonda-fbk** [eclipse-xtext/xtext#3151] [Q&A] [answered]
Hi, while inspecting the Xtext grammar for Xtext itself, I've noticed that rules support parameters.

```
fragment RuleNameAndParams returns ParserRule:
        name=ValidID ('<' (parameters+=Parameter (',' parameters+=Parameter)*)? '>')?
;
```

It seems that these parameters can take as values boolean literals, references to other parameters, or combinations thereof.

However, I could find zero documentation online that explains when rule parameters could be useful (i.e. examples or design patterns), and, most importantly, how to use them (i.e. perhaps access them programmatically and do something based on their values? I'm not sure, I am just trying to imagine the main use case). Could anyone enlighten me, please?

Cheers :)

**Comment by cdietrich** [comment, marked as answer]
I fear there is no such thing like docs for this. You might find some examples in the xtext repo and some other repos eg n4js 

it basically allows to to turn of parts of rules from outside 
So that you can use the same rule at multiple places and then turn of some alternatives at different places

**Reply by tfonda-fbk** [reply]
@cdietrich 
Thanks for pointing me to the n4js project. I gave a look at it and quickly found a usage scenario which helped me understand. In particular, it was unclear to me how to use a parameter's value inside a rule to make a parsing choice, but I got it now. So I have created a very stupid example to prove the correctness of my understanding:

```
Model:
	(members+=Entity<true> | members+=Block)*;

Block:
	{Block} '{' members+=Entity<false>* '}';

Entity<param>:
	(<param> 'entity' | <!param> 'Entity') name=ID;
```

For other readers: this grammar requires a capitalized `Entity` keyword if an entity is declared inside a block, since in rule Block Entity is called with its parameter set to false. Outside of blocks, the `entity` keyword must be all lowercase.

I will check if in my project there are spots rule parameters could be of help in (I'm pretty sure there are). In the next few weeks, I will surely contribute some documentation about them.

**Comment by cdietrich** [comment]
You basically for a rule define 

<a,b,c>

In the alternative add a 

<a> Alternative1
|
<b> Alternative2

And then call rules with <a=true> or passing from parent
